diag.c

```diff
@@ -0,0 +1,225 @@
+#include "git-compat-util.h"
+#include "json-writer.h"
+#include "version.h"
+#include "thread-utils.h"
+
+#ifndef NO_CURL
+#include <curl/curl.h>
+#endif
+
+/*
+ * Fixed details of the compiled architecture.
+ */
+static void cmd__arch(struct json_writer *jw)
+{
+	/*
+	 * We do not probe the system for the actual CPU architecture name.
+	 * We just report the value of the #define from the Makefile.
+	 */
+	jw_object_string(jw, "defined", GIT_HOST_CPU);
+
+	jw_object_intmax(jw, "sizeof_long", sizeof(long));
+	jw_object_intmax(jw, "sizeof_size_t", sizeof(size_t));
+	jw_object_intmax(jw, "sizeof_intmax_t", sizeof(intmax_t));
+	jw_object_intmax(jw, "sizeof_pointer", sizeof(void *));
+}
+
+#ifdef GIT_WINDOWS_NATIVE
+static void system_helper_windows(struct json_writer *jw)
+{
+	SYSTEM_INFO info;
+	GetSystemInfo(&info);
+
+	jw_object_inline_begin_object(jw, "windows");
+
+	jw_object_intmax(jw, "processor_architecture",
+			 info.wProcessorArchitecture);
+	jw_object_intmax(jw, "processor_level", info.wProcessorLevel);
+	jw_object_intmax(jw, "processor_revision", info.wProcessorRevision);
+
+	jw_end(jw);
+}
+#endif
+
+/*
+ * Details of the actual running system.
+ */
+static void cmd__system(struct json_writer *jw)
+{
+	jw_object_intmax(jw, "online_cpus", online_cpus());
+
+#ifdef GIT_WINDOWS_NATIVE
+	system_helper_windows(jw);
+#endif
+}
+
+/*
+ * Details of the running version of Git.
+ */
+static void cmd__version(struct json_writer *jw)
+{
+	jw_object_string(jw, "git", git_version_string);
+
+	jw_object_bool(jw, "built_from_commit", git_built_from_commit_string[0]);
+	if (git_built_from_commit_string[0])
+		jw_object_string(jw, "commit", git_built_from_commit_string);
+}
+
+static void do_string_or_null(struct json_writer *jw, const char *key,
```

Comment:
This function seems to be used only in the `!NO_CURL` part, and should therefore be guarded by `#ifndef NO_CURL ... #endif`. That's what makes the Linux/macOS builds fail, as far as I can tell.